### ISSUE TITLE
On Windows emsdk_env.bat removes most of the PATH environment variable

### ISSUE THREAD

**Issue by yowl** [emscripten-core/emsdk#634]
Checkout commit with hash 807e55cbec8fec292045ebcaf547ed7466762d7f, start a command shell:
```
E:\>cd github

E:\GitHub>emsdk\emsdk_env.bat
The system cannot find the drive specified.
Adding directories to PATH:
PATH += E:\GitHub\emsdk
PATH += E:\GitHub\emsdk\node\12.18.1_64bit\bin
PATH += E:\GitHub\emsdk\python\3.7.4-pywin32_64bit
PATH += E:\GitHub\emsdk\java\8.152_64bit\bin

Setting environment variables:
EMSDK = E:/GitHub/emsdk
EM_CONFIG = E:\GitHub\emsdk\.emscripten
EMSDK_NODE = E:\GitHub\emsdk\node\12.18.1_64bit\bin\node.exe
EMSDK_PYTHON = E:\GitHub\emsdk\python\3.7.4-pywin32_64bit\python.exe
JAVA_HOME = E:\GitHub\emsdk\java\8.152_64bit

E:\GitHub>powershell
'powershell' is not recognized as an internal or external command,
operable program or batch file.

E:\GitHub>echo %PATH%
E:\GitHub\emsdk;E:\GitHub\emsdk\node\12.18.1_64bit\bin;E:\GitHub\emsdk\python\3.7.4-pywin32_64bit;E:\GitHub\emsdk\java\8.152_64bit\bin;C:\python3;C:\Users\xxxxx\AppData\Local\Microsoft\WindowsApps;C:\Users\xxxxx\.dotnet\tools;C:\Program Files\Wasmtime\bin;C:\Users\xxxxx\AppData\Local\Programs\Fiddler;C:\Program Files (x86)\Sophos\Sophos SSL VPN Client\bin;C:\Users\xxxxx\AppData\Roaming\npm;C:\Users\xxxxx\AppData\Local\Microsoft\WindowsApps;C:\Users\xxxxx\.dotnet\tools

E:\GitHub>
```
The PATH has been pretty much replaced.  Before running `emsdk_env.bat`  PATH is much longer, fills the screen.

Check out emsdk before this commit #621  and it is fine.

**Comment by sbc100** [comment]
@aminya can you take a look?

**Comment by aminya** [comment]
I don't quite understand the issue. What do you mean by it removes most of the PATH environment variable?

**Comment by sbc100** [comment]
If you do `echo %PATH%` before and after it looks like it removes a bunch of paths as well as adding the emsdk ones.  

Maybe PATH on windows is constructed based on the sum of system and user values for PATH in the registry..?

**Comment by aminya** [comment]
I can look into this, but emsdk code for adding environment variables is very long and hard to follow. I don't understand why we don't simply use a basic PowerShell script for doing that.

**Comment by sbc100** [comment]
Maybe we are conflating.. "what path should I store in the registry" with "what PATH do we want in the current shell".

For the current shell we always want to add to the current `$PATH` .. and we probably don't want to look in the registry at all.

**Comment by sbc100** [comment]
i.e. unless you specify `--persistent` we should probably never look in the registry should we?

**Comment by yowl** [comment]
I think line 2490 should be 
```
existing_path = win_get_environment_variable('PATH', system==system).split(ENVPATH_SEPARATOR)
```

Double == not =

**Comment by yowl** [comment]
Sorry, mean https://github.com/emscripten-core/emsdk/blob/31f5f842ff4c8d22e939fe0c8c27b34d9c9c0184/emsdk.py#L2488

**Comment by aminya** [comment]
@yowl Could you try https://github.com/emscripten-core/emsdk/pull/515?

**Comment by aminya** [comment]
> i.e. unless you specify `--persistent` we should probably never look in the registry should we?

On Windows, these two are the same. PATH is only defined in the registry and it is available in the shell.

If users do not use --persistent it should not touch the registry.

**Comment by sbc100** [comment]
> I think line 2490 should be
> 
> ```
> existing_path = win_get_environment_variable('PATH', system==system).split(ENVPATH_SEPARATOR)
> ```
> 
> Double == not =

I don't think that is it.    `system==system` is an expression that is the same as `true`

**Comment by yowl** [comment]
@aminya #515 fixes it.  I see it removes the line with the single `=`

**Comment by sbc100** [comment]
I think the problem is that we always use `win_get_environment_variable` to get the current PATH on windows but we really need to use `os.envrion['PATH']` when modify the PATH in the existing shell.. othewise we clobber any local changes to %PATH% made by the user in the current shell.

**Comment by yowl** [comment]
> 
> 
> > I think line 2490 should be
> > ```
> > existing_path = win_get_environment_variable('PATH', system==system).split(ENVPATH_SEPARATOR)
> > ```
> > 
> > 
> > Double == not =
> 
> I don't think that is it. `system==system` is an expression that is the same as `true`

ok, my python is no good:-)

**Comment by yowl** [comment]
> 
> 
> I think the problem is that we always use `win_get_environment_variable` to get the current PATH on windows but we really need to use `os.envrion['PATH']` when modify the PATH in the existing shell.. othewise we clobber any local changes to %PATH% made by the user in the current shell.

that sounds likely, as I run emscripten from a developer shell which sets up other stuff

**Comment by fibbo** [comment]
When running `emsdk activate` the previous system `PATH` environment should imo be preserved and not be dependent on any `--system` parameter.

https://github.com/emscripten-core/emsdk/blob/53c184e8d3b82c2cc9c03dc8aeffe623dd29b4a9/emsdk.py#L2455

used to have `system=True` which then would work properly. Is there anything that would speak against reverting this?

**Comment by aminya** [comment]
> When running `emsdk activate` the previous system `PATH` environment should imo be preserved and not be dependent on any `--system` parameter.
> 
> https://github.com/emscripten-core/emsdk/blob/53c184e8d3b82c2cc9c03dc8aeffe623dd29b4a9/emsdk.py#L2455
> 
> used to have `system=True` which then would work properly. Is there anything that would speak against reverting this?

That is irrelevant. `System=True` means that emsdk will try to access Machine environment variables instead of User.

This PR will probably solve this issue.
https://github.com/emscripten-core/emsdk/pull/515

**Comment by Squareys** [comment]
@aminya This issue is still present

**Comment by Squareys** [comment]
@fibbo Aparently setting `emsdk_env.bat --system` will pass `--system` to emsdk.py and set that to `True` there.
I guess by default this should be on 🤔

**Comment by aminya** [comment]
@Squareys Could you try #515?

**Comment by sbc100** [comment]
I don't really see how #515 could solve the issue since we still *read* the exist environment variables from the registry..  the ---permanent argument only seems to have en effect on the *write* part, not the read part, unless I'm missing something?

 I think we would need a mode that reads from the current environment if we want to avoid clobbering it?

**Comment by aminya** [comment]
The Master branch is updated. Can people try again and report if it resolves the issue?

**Comment by Squareys** [comment]
@aminya It worked via git shell (msys), but not via cmd.

**Comment by yowl** [comment]
@aminya not working for me via cmd either.  PATH is clobbered

**Comment by yowl** [comment]
In fact, looking closer, with #515 from kudaba's branch, the path is also clobbered but not as much.  I hadn't noticed before as the part of the path that I needed were still there.  With the master branch, more of the path is lost and I loose for example, powershell, devenv.exe.

**Comment by Squareys** [comment]
```
diff --git a/emsdk.py b/emsdk.py
index 5032244..f419781 100644
--- a/emsdk.py
+++ b/emsdk.py
@@ -2500,7 +2500,7 @@ def to_msys_path(p):

 # Looks at the current PATH and adds and removes entries so that the PATH reflects
 # the set of given active tools.
-def adjusted_path(tools_to_activate, system=False):
+def adjusted_path(tools_to_activate, system=True):
   # These directories should be added to PATH
   path_add = get_required_path(tools_to_activate)
   # These already exist.
```

Fixes the issue (partially, see below).

Only the user's `PATH` was retrieved, and then the global `PATH` was set to that plus the emsdk paths.
That means that the global/system `PATH` parts are lost, they are not inherited when retrieving the path with `system=False`.

I assume this was broken while adding the ability to make the path permanent: If setting for user only, you want to retrieve only the User parts of `PATH`, otherwise you'd be copying all of the global path into the user path.

Interestingly, only the System part is retrieved with `system=True`, so setting the non-permanent path should probably be done by joining new, user and system parts.

**Comment by aminya** [comment]
This does not solve the issue. It circumvents it. We need to look at the `adjusted_path` to see why it cannot split and add PATH correctly!
https://github.com/emscripten-core/emsdk/blob/da9419f42ad3ed112adedbb830ef5140f92af4b6/emsdk.py#L2503-L2529

**Comment by aminya** [comment]
Could you check https://github.com/emscripten-core/emsdk/pull/642/ and see if it solves the issue for you?

**Comment by aminya** [comment]
I am totally sure now that this is fixed in #642. 🎉  Adding the tests helped me to find the issue.

**Comment by sbc100** [comment]
The exact part of #621 that causes this breakage seems to be the removal of the contantication of User and System paths:

```
-    existing_path = win_get_environment_variable('PATH', system=True)
-    if not system_path_only:
-      current_user_path = win_get_environment_variable('PATH', system=False)
-      if current_user_path:
-        existing_path += ENVPATH_SEPARATOR + current_user_path
-    existing_path = existing_path.split(ENVPATH_SEPARATOR)
```

The old code would take both the system and user paths and add them together.

**Comment by aminya** [comment]
> The exact part of #621 that causes this breakage seems to be the removal of the contantication of User and System paths:

This is already fixed in #642. I am not sure what we are waiting for.
